DOC: Add documentation for Tensor.is_nonzero

bharatr21 opened the pull request:
Fixes #37438 by adding documentation for `Tensor.is_nonzero`

bharatr21 committed:
DOC: Add documentation for Tensor.is_nonzero

bharatr21 committed:
DOC: Add documentation for Tensor.is_nonzero

vadimkantorov commented:
Does this method work for sparse tensors? (coalesced and non-coalesced, just emtpy thus zero ones?) it would be nice for docs to be explicit about this

Bharat123rox commented:
> Does this method work for sparse tensors? (coalesced and non-coalesced, just emtpy thus zero ones?) it would be nice for docs to be explicit about this

This method simply throws a `RuntimeError` if `numel() < 1` so it should give an error. I'll add that example though

bharatr21 committed:
DOC: Add info for sparse tensors

vadimkantorov commented on torch/_torch_docs.py:
```diff
@@ -2703,7 +2703,8 @@ def merge_dicts(*dicts):
 i.e. not equal to ``torch.Tensor([0.])`` or ``torch.Tensor([0])`` or
 ``torch.Tensor([False])``.
-Throws a ``RuntimeError`` if ``torch.numel() != 1``.
+Throws a ``RuntimeError`` if ``torch.numel() != 1`` (after coalescing in case
```
Does "after coalescing" mean that `is_nonzero`  coalesces every sparse tensor?

Bharat123rox replied:
No but `numel()` is calculated taking the coalesced tensor / dense tensor into account, this is not an inplace operation so it will not modify the tensor

Bharat123rox replied:
I'm not getting the right phrase for that

Commit:
[PyTorch Numeric Suite] Add module output comparison (#36701)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/36701

Add module output comparison API.
ghstack-source-id: 103368194

Test Plan: buck test mode/dev caffe2/test:quantization -- 'test_compare_model_outputs'

Differential Revision: D21053197

fbshipit-source-id: cabcafbeeac1b604db069833a0f17ebce506ba65

Commit:
[caffe2] L2 regularization for rowwise fused sparse adagrad (#37653)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37653

Following up D21320243 adding weight_decay to rowwise fused sparse adagrad. This is more involved because we can't reuse g_sq_avg multiple times.

Test Plan: CI

Reviewed By: jspark1105

Differential Revision: D21335643

fbshipit-source-id: 491b385c5eb9c0d1e3d31a1cf50d7eb450c2d39d

Commit:
implement L2 regularization for Adagrad in caffe2 and dper (#37705)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37705

Pull Request resolved: https://github.com/pytorch/pytorch/pull/37372

Posted note: [Regularizing SparseNN Against Over-fitting](https://fb.workplace.com/notes/taiqing-wang/regularizing-sparsenn-against-over-fitting/220306075902708/)

**Problem formulation**

L(w) = J(w) + lambda/2 * ||w||^2
J(w) is the empirical loss, and ||w||^2 is the squared L2 norm of the parameters, a.k.a. L2 regularizer.

dL(w)/ dw_i = dJ(w)/dw_i + lambda w_i
dL(w)/ dw_i is the gradient of L(w) w.r.t. w_i.

To implement the L2 regularizer, the gradient of J(w) w.r.t. w_i is added with w_i. lambda is called as weight decay in this implementation.

**Code changes**
* In the initialization method of AdagradOptimizer, a new input argument, weight_decay, is added.
* In the _run function of AdagradOptimizer, the weight decay will be skipped for 1d bias vectors.
* In the parameter update functions of Adagrad, the gradient is updated by weight_decay * w_i. The default value for weight_decay is zero.

Test Plan:
`
buck build caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_weight_decay
`

`
./buck-out/gen/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_weight_decay#binary.par
`

Reviewed By: jspark1105

Differential Revision: D21258652

fbshipit-source-id: d2366ddcd736a03205a2d16f914703b16d9fce8f

Commit:
[caffe2] L2 regularization for fused sparse Adagrad (#37652)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37652

Add weight_decay to fused adagrad operators. This should be landed with the next diff together. Just separating out to make review easier.

Test Plan: CI

Reviewed By: jspark1105

Differential Revision: D21320243

fbshipit-source-id: 1157471988dedd60ba9b62949055f651b1fa028f

Commit:
Updating submodules

Summary:
GitHub commits:

https://github.com/pytorch/fbgemm/commit/fb6b95d8c7efadb3ce56bc23a5391b29be216e64

Test Plan: n/a

Reviewed By: yns88

fbshipit-source-id: 9ff6b68322b92710c7356030bc9e4c364a937ea3

linbinyu committed:
Add ops for portal NLU model (#37192)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37192

Add ops used by portal NLU model to lite interpreter

Test Plan: local test

Reviewed By: iseeyuan

Differential Revision: D21215050

fbshipit-source-id: 874023c449e4c04b9f3f871450a7cf02e8f5f5c4

Commit:
[caffe2] Shape inference for UnPackRecords

Summary:
Since UnPackRecords is part of the graph, we need to add shape inference for it to make it work e2e with tvm_jit_op. Because the input is packed, shape inference is impossible without shape info of the packed tensors. Some context, the shape of the packed tensor is 1 X num_embeddings X embedding_size, with 1 being the batch_size. The shape of the corresponding output tensor is thus batch_size X num_embeddings X embedding_size after concatenating the packed tensors on the batch axis. Therefore two more gflags need to be added

- caffe2_predictor_num_embeddings
- caffe2_predictor_embedding_size

These gflags are then added to the UnPackRecordsOp in the predict_net as args to pass the info to c2_frontend so TVM can do its own shape inference.

Reviewed By: yinghai

Differential Revision: D21286983

fbshipit-source-id: e9a19cb6b564905282a771df2b9d211d5d37dd71

Commit:
[TVM] Implement UnPackRecordsOp (#37489)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37489

Reviewed By: yinghai

Differential Revision: D21196596

fbshipit-source-id: 58b8bc3afc472e02ef7e3d31151fe8ace2be2a73

lw committed:
Update TensorPipe submodule (#37729)

Summary:
In order to include these fixes that were blocking https://github.com/pytorch/pytorch/pull/35483:
- https://github.com/pytorch/tensorpipe/commit/673eda9efc167f82879fea8245d9d432937833c8
- https://github.com/pytorch/tensorpipe/commit/ff8d1733ad861ad3eaa3e293a288a8eeb4fc3d19
- https://github.com/pytorch/tensorpipe/commit/c73367836f27b92e5808de390915f6fccae9a991

Pull Request resolved: https://github.com/pytorch/pytorch/pull/37729

Reviewed By: beauby

Differential Revision: D21378972

Pulled By: lw

fbshipit-source-id: 3375fe1fa6e79817da3bb033127c3c8f31c3ffc3

ezyang committed:
Revert D21215050: Add ops for portal NLU model

Test Plan: revert-hammer

Differential Revision:
D21215050

Original commit changeset: 874023c449e4

fbshipit-source-id: 695494d9607bc8823c494fa06830370adccbf935

mrityunjay-tripathi committed:
typo in MultiheadAttention documentation (#37496)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37496

Differential Revision: D21356999

Pulled By: zou3519

fbshipit-source-id: ba6c7b19053c97e5ab77eb0c44e97d26b04d85da

Commit:
Fix for tests on ROCm (#37616)

Summary:
This pull request fixes and re-enables two of the tests disabled in https://github.com/pytorch/pytorch/issues/37427
1. `test_sparse_add_out_bfloat16` in test_sparse.py fixed to use updated `atol` argument instead of `prec` for `assertEqual`
2. The conversion of `flt_min` to `int64` is divergent on HIP compared to numpy. The change removes that conversion from the `test_float_to_int_conversion_finite` test case in test_torch.py

cc: ezyang jeffdaily
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37616

Differential Revision: D21379876

Pulled By: ezyang

fbshipit-source-id: 2bfb41d67874383a01330c5d540ee516b3b07dcc

ShawnZhong committed:
[CircleCI] Store build artifacts for python docs (#37658)

Summary:
This PR allows the build artifacts for python docs to be stored on CircieCI, which helps the reviewer to preview doc changes before merging.

The artifacts can be found in the [`ARTIFACTS` tab](  https://app.circleci.com/pipelines/github/pytorch/pytorch/162986/workflows/a969f256-3243-414f-8a02-1234b9dac149/jobs/5320907/artifacts) of the test **pytorch_cpp_doc_push**, and the website is served at https://5320907-65600975-gh.circle-artifacts.com/0/docs/index.html

This PR is inspired by rgommers's comment under https://github.com/pytorch/pytorch/pull/37419#issuecomment-621420500

> There's a CircleCI job pytorch_python_doc_push that builds the docs, however it doesn't store any artifacts for PRs. Controlled by .circleci/scripts/python_doc_push_script.sh. I think that's the only doc build (?). Not sure why it doesn't store the artifacts, that would be useful.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37658

Differential Revision: D21380094

Pulled By: ezyang

fbshipit-source-id: 1dd44bf836ebc74454f4444ae9321807dccdb313

gchanan committed:
Move baddbmm broadcasting from codegen layer to native layer. (#37544)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37544

Test Plan: Imported from OSS

Differential Revision: D21315039

Pulled By: gchanan

fbshipit-source-id: aa564d06d415ad2468b4898f84edbb03a3ee698f

Baranowski committed:
Remove zero_dim_dispatch_when_scalar (#37580)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/33094
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37580

Differential Revision: D21380154

Pulled By: ezyang

fbshipit-source-id: 4556c7ca6126a7d382f6343aee14a7e46c498ac3

mattip committed:
Split up documentation into subpages and clean up some warnings (#37419)

Summary:
xref gh-32838, gh-34032

This is a major refactor of parts of the documentation to split it up using sphinx's `autosummary` feature which will build out `autofuction` and `autoclass` stub files and link to them. The end result is that the top module pages like torch.nn.rst and torch.rst are now more like table-of-contents to the actual single-class or single-function documentations pages.

Along the way, I modified many of the docstrings to eliminate sphinx warnings when building. I think the only thing I changed from a non-documentation perspective is to add names to `__all__` when adding them to `globals()` in `torch.__init__.py`

I do not know the CI system: are the documentation build artifacts available after the build, so reviewers can preview before merging?
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37419

Differential Revision: D21337640

Pulled By: ezyang

fbshipit-source-id: d4ad198780c3ae7a96a9f22651e00ff2d31a0c0f

Commit:
LoopOptions::gpu_block_index(): bool -> int (#37578)

Summary:
Small change to allow MSVC build pass.
The error is

```
D:\pytorch-scripts\caffe2_builders\v141\pytorch\torch/csrc/jit/tensorexpr/stmt.h(370): error C4805: '!=': unsafe mix
of type 'bool' and type 'int' in operation (compiling source file D:\pytorch-scripts\caffe2_builders\v141\pytorch\torch
\csrc\jit\passes\tensorexpr_fuser.cpp) [D:\pytorch-scripts\caffe2_builders\v141\pytorch\build\RelWithDebInfo\caffe2\tor
ch_cpu.vcxproj]
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37578

Differential Revision: D21348964

Pulled By: ezyang

fbshipit-source-id: 2c5f995e0adbeb681c18625b59250d7ee3e958ef

jspark1105 committed:
[caffe2/dnnlowp] documentation for pack operator arguments (#37719)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37719

As title

Test Plan: Just updating doc

Reviewed By: hyuen

Differential Revision: D21369227

fbshipit-source-id: a45e5d0fa34aea8046eb4bb83e6c4df4d2654252

suo committed:
add fmt to libtorch_python.so (#37560)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37560

Test Plan: Imported from OSS

Differential Revision: D21320059

Pulled By: suo

fbshipit-source-id: 95cfe7cf26c515fdfcb4621cc58266d838a38a3e

Commit:
Remove unnecessary pickle and unpickle invocation in PyRRef __setstate__/__getstate__ methods (#37638)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37638

Test Plan: Imported from OSS

Differential Revision: D21343280

Pulled By: mrshenli

fbshipit-source-id: da462fee5815dc74c7f2dc3161699e461bc7d7d3

zasdfgbnm committed:
Remove std::complex from TypeMeta (#37632)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37632

Differential Revision: D21362056

Pulled By: anjali411

fbshipit-source-id: b20506a36594ad8485ba8ef31d2d8a83ff0862f2

pbelevich committed:
Show warning if Tensor.random_()'s from and to are not in [-(2^digits), 2^digits] bounds for floating-point types (#37537)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37537

The documentation states that `random_()` samples "from the discrete uniform distribution". Floating-point types can support _discrete_ _uniform_ distribution only within range [-(2^digits), 2^digits], where `digits = std::numeric_limits<fp_type>::digits`, or

- [-(2^53), 2^53] for double
- [-(2^24), 2^24] for double
- [-(2^11), 2^11] for half
- [-(2^8), 2^8] for bfloat16

The worst scenario is when the floating-point type can not represent numbers between `from` and `to`. E.g.
```
torch.empty(10, dtype=torch.float).random_(16777217, 16777218)
tensor([16777216., 16777216., 16777216., 16777216., 16777216., 16777216.,
        16777216., 16777216., 16777216., 16777216.])
```
Because 16777217 can not be represented in float

Test Plan: Imported from OSS

Differential Revision: D21380387

Pulled By: pbelevich

fbshipit-source-id: 80d77a5b592fff9ab35155a63045b71dcc8db2fd

jianyuh committed:
[TS][easy] Typo Fix (#37773)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37773

As Title says
ghstack-source-id: 103385174

Test Plan: CI

Reviewed By: dmudiger

Differential Revision: D21374951

fbshipit-source-id: a2fc48b931f0cecbc8a995bf4b4ace30a8eb0d70

suo committed:
Revert D21337640: [pytorch][PR] Split up documentation into subpages and clean up some warnings

Test Plan: revert-hammer

Differential Revision:
D21337640

Original commit changeset: d4ad198780c3

fbshipit-source-id: fa9ba6ac542173a50bdb45bfa12f3fec0ed704fb

malfet committed:
[PyTorch] Modify `data_parallel` to work with small tensors (#37704)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37704

If input tensor can not be chunked, run `parallel_apply` on fewer devices
Modfy input tensor dimention in `DataParallelUsesAllAvailableCUDADevices_CUDA` to be chunkable by any number of available CUDA devices

Test Plan: Run `test/cpp/api/parallel` on machine  with 6 GPUs

Differential Revision: D21365416

fbshipit-source-id: 60fdfed4a0e6256b2c966c2ea3e8d0bfb298d9a8

zasdfgbnm committed:
Migrate item() to c10::complex (#37648)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37648

Test Plan: Imported from OSS

Differential Revision: D21382318

Pulled By: anjali411

fbshipit-source-id: c1d3da43f118f18739bb34906f76a5bad097c905

mpjlu committed:
To fix caffe2 model with Copy OP cannot export to onnx model (#37144)

Summary:
To fix caffe2 model with Copy OP cannot export to onnx model
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37144

Reviewed By: houseroad

Differential Revision: D21252421

Pulled By: yinghai

fbshipit-source-id: 4f1077188f36b0691d199e418880bbb27f11032d

ezyang committed:
Make msg() and msg_with_backtrace() private (#37094)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37094

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D21202892

Pulled By: ezyang

fbshipit-source-id: d59e6bffabd90cc734056bdce2cd1fe63262fab8

ezyang committed:
Refactor error msg stack handling, add TORCH_RETHROW (#37101)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37101

Fixes #36954.

The basic concept is to streamline the process of rethrowing
c10::Error with extra error information.  This is in a few
steps:

- I completely remodeled the Error data type and the internal
  invariants.  Instead of manually adding in newlines, the
  message stack formatting process is responsible for inserting
  newlines and spacing as necessary.  Call sites are then
  modified to respect the new API model.
- TORCH_RETHROW macro is added, which adds context to an error
  message and then rethrows it.

New internal assert failure looks like:

```
0 INTERNAL ASSERT FAILED at ../c10/test/util/exception_test.cpp:64, please report a bug to PyTorch.
Exception raised from TestBody at ../c10/test/util/exception_test.cpp:64 (most recent call first):
frame #0: <unknown function> + 0x6aab9 (0x7ff611d3aab9 in /data/users/ezyang/pytorch-tmp/build/lib/libc10.so)
frame #1: ...
```

Error message with context looks like:

```
This is an error
  This is context 1
  This is context 2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D21202891

Pulled By: ezyang

fbshipit-source-id: 361cadd16bc52e5886dba08e79277771ada76169

bwasti committed:
Permit registration of multiple triggers, but insert warning (#37772)

Summary:
If linking the same file multiple times, the trigger check becomes severe and crashes execution at startup.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37772

Differential Revision: D21384072

Pulled By: bwasti

fbshipit-source-id: 3396e69cd361f65e50517970d23497804c76023e

supriyar committed:
[quant][graph] Fix bug in replicateDequant (#37637)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37637

Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Imported from OSS

Differential Revision: D21354931

fbshipit-source-id: 79a1dc63b0ed96c3d51d569116ed963106085d3b

supriyar committed:
[quant][graph] Fix bug in replaceConvolutionWithConv2d (#37635)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37635

replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan: Imported from OSS

Differential Revision: D21354930

fbshipit-source-id: a2dbad856666b4bbb2d9015ade8e1704774f20dd

Krovatkin committed:
capitalize fuseTensorExpr (#37780)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37780

Differential Revision: D21386092

Pulled By: Krovatkin

fbshipit-source-id: c190f891fe25b3cee9a34b5173756c39efd49c66

gchanan committed:
Kill the ability to codegen tensor-based broadcasting. (#37547)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37547

This shouldn't be used anymore.

Test Plan: Imported from OSS

Differential Revision: D21315037

Pulled By: gchanan

fbshipit-source-id: 12728f1d0e1856bf3e8fe1bfcf36cddd305a4a76

gchanan committed:
Move broadcasting code for fmod, fmod_ from codegen layer. (#37545)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37545

Test Plan: Imported from OSS

Differential Revision: D21315036

Pulled By: gchanan

fbshipit-source-id: cbe82205dc71c2a704d717a5f82827fc6ff5106c

gchanan committed:
Move addr broadcasting from codegen layer to native layer. (#37546)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37546

Test Plan: Imported from OSS

Differential Revision: D21315040

Pulled By: gchanan

fbshipit-source-id: 1bba97bd889ec286e3e7f1d0f0450871b996c9ae

ezyang committed:
Revert "Revert D21337640: [pytorch][PR] Split up documentation into subpages and clean up some warnings" (#37778)

Summary:
Original PR: https://github.com/pytorch/pytorch/pull/37419

cc mattip suo
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37778

Differential Revision: D21385774

Pulled By: ezyang

fbshipit-source-id: 5de532faab8bae132736b6b5189e0ee2ac9935be

supriyar committed:
[quant][mobile] Return for conv with empty batch (#37779)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37779

We should just return empty output

Test Plan: Imported from OSS

Differential Revision: D21385789

fbshipit-source-id: 4b42f5aaebabfa3f329ed74356bddb33daad98d5

zasdfgbnm committed:
Migrate CPU casting copy kernel to c10::complex (#37649)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37649

Test Plan: Imported from OSS

Differential Revision: D21385430

Pulled By: anjali411

fbshipit-source-id: 66106edfc682fc75f293babece1dc4323aa3aeb1

rohan-varma committed:
[MultiProcessTestCase] Improve the error message when a process terminates (#37627)

Summary:
When a subprocess terminates with an exception in a distributed test, log the process number as well
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37627

Differential Revision: D21366149

Pulled By: rohan-varma

fbshipit-source-id: 132c4b4c1eb336761c2be26d034d8b739ae19691

suo committed:
split out docs tests into separate job (#37793)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37793

Test Plan: Imported from OSS

Differential Revision: D21392798

Pulled By: suo

fbshipit-source-id: 172fb0522d0b168ca19a382e5fb1eb87b6390acc

p12tic committed:
[ONNX] Fix numerical errors in softmax when dim is not last dimension (#37326)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/34585.

This PR improves the workaround for the problem of different semantics between ONNX softmax and Pytorch softmax.

In Pytorch the `dim` parameter specifies over which dimension normalize the values.  ONNX on the other hand always coerces the input into a 2D tensor and the `axis` parameter specifies which dimensions represent rows and columns of the resulting tensor. As a result, only when we are normalizing the last dimension (`dim == ndim - 1`) semantics are the same.

Previously this was handled by recognizing the `dim == ndim - 1` case and using `softmax` for that. All other cases used a fallback path of explicit invocations of exp, reducesum and div operators to compute the result. Unfortunately, this results in numeric errors when input values are large: the result of exp will produce infinity on both numerator and denumerator and the division of that will result in NaN.

This can be improved by transposing the input tensor so that we can reuse ONNX softmax.

Similar approach has been applied to `logsoftmax` function in https://github.com/pytorch/pytorch/issues/30433.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37326

Reviewed By: hl475

Differential Revision: D21389712

Pulled By: houseroad

fbshipit-source-id: 554fd1b98231a28984c30c7e7abd3c0643386ff7

bharatr21 committed:
Update README to include few (missing?) links (#37714)

Summary:
Update of README
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37714

Differential Revision: D21393786

Pulled By: ngimel

fbshipit-source-id: 8ae12b38989cbfcdd4d69db1c1ab3bbac0e0db61

ngimel commented:
Can you please resolve conflicts?

Commit:
Move device type init from BackendSelect to backend kernels (#37402)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37402

Previously, BackendSelect kernels did just-in-time device type
initialization by calling `LegacyTypeDispatch.initForDispatchKey()`
with a computed dispatch key. Here we move the initialization to
the backend kernels themselves, where we can call the device-
specific initializer directly.

Putting this up to run tests on it, but a couple questions remain:
* why were only BackendSelect kernels doing this initialization?
  Not all factory ops appear there, nor are all the ops that do
  appear there factory ops. Currently we generate init code for
  exactly the BackendSelect ops, but the choice should be better
  motivated.
* the previous scheme maps HIP to its own legacy type dispatch
  entry, but the logic assumes it's exclusive with CUDA, and no
  ops appear to mention HIP explicitly, so the new logic doesn't
  expose a static entry point for it. Needs to be verified.

Test Plan: Imported from OSS

Differential Revision: D21282974

Pulled By: bhosmer

fbshipit-source-id: cd46eb788596948e0572a15fac0f8b43feca5d75

zasdfgbnm committed:
Migrate CUDA cat, scatter, gather, index, index_put to c10::complex (#37650)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37650

Test Plan: Imported from OSS

Differential Revision: D21394299

Pulled By: anjali411

fbshipit-source-id: dd7666af736e720b54b978cc62570d0e840e092f

frgfm committed:
docs: Fixed docstring indentation for documentation (#37739)

Summary:
Hello there,

I was going through the default initialization of some layers, and ended up on the `torch.nn.init` documentation. As shown below, there was a slight issue with the docstrings of both `kaiming_normal_` and `kaiming_uniform_` that yielded a wrong list of function parameters:

![doc_issue](https://user-images.githubusercontent.com/26927750/80923512-88e30400-8d84-11ea-8708-36ed3a0f7749.png)

This PR fixes the indentation in the corresponding docstrings.

Any feedback is welcome!
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37739

Differential Revision: D21393728

Pulled By: ngimel

fbshipit-source-id: 64523cb328e72d2e51c2c42b20a4545c1ec5f478

crcrpar committed:
Use `gpu_kernel` in Affine Quantizer (#37312)

Summary:
Removes `CUDA_tensor_apply2` from Affine Quantizer.

cc: zasdfgbnm

# Profiling

## This PR

### quint8

```==4458==       Range "quantize_per_tensor, seq = 0"
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
          Range:  100.00%  4.8703ms        20  243.52us  207.60us  312.66us  quantize_per_tensor, seq = 0
 GPU activities:  100.00%  751.95us        10  75.194us  74.372us  79.044us  _ZN2at6native6modern29vectorized_elementwise_kernelILi4EZZZNS0_75_GLOBAL__N__51_tmpxft_0000424b_00000000_6_affine_quantizer_cpp1_ii_92f2f7d738quantize_tensor_per_tensor_affine_cudaENS_6TensorES4_dlENKUlvE_clEvENKUlvE0_clEvEUlfN3c106quint8EE_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_
      API calls:  100.00%  162.48us        10  16.247us  13.383us  35.997us  cudaLaunchKernel
```

### qint8

```==14289==       Range "quantize_per_tensor, seq = 0"
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
          Range:  100.00%  4.8143ms        20  240.71us  155.68us  327.78us  quantize_per_tensor, seq = 0
 GPU activities:  100.00%  748.85us        10  74.884us  73.892us  78.565us  _ZN2at6native6modern29vectorized_elementwise_kernelILi4EZZZNS0_75_GLOBAL__N__51_tmpxft_0000424b_00000000_6_affine_quantizer_cpp1_ii_92f2f7d738quantize_tensor_per_tensor_affine_cudaENS_6TensorES4_dlENKUlvE_clEvENKUlvE_clEvEUlfN3c105qint8EE_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_
      API calls:  100.00%  166.61us        10  16.661us  13.387us  39.237us  cudaLaunchKernel
```

### qint32

```
==17303==       Range "quantize_per_tensor, seq = 0"
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
          Range:  100.00%  19.011ms        20  950.55us  308.07us  1.0331ms  quantize_per_tensor, seq = 0
 GPU activities:  100.00%  1.1440ms        10  114.40us  113.42us  117.74us  _ZN2at6native6modern29vectorized_elementwise_kernelILi4EZZZNS0_75_GLOBAL__N__51_tmpxft_0000424b_00000000_6_affine_quantizer_cpp1_ii_92f2f7d738quantize_tensor_per_tensor_affine_cudaENS_6TensorES4_dlENKUlvE_clEvENKUlvE1_clEvEUlfN3c106qint32EE_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_
      API calls:  100.00%  163.78us        10  16.378us  13.747us  35.668us  cudaLaunchKernel
```

## Original

commit: b428f454e13f6e8055124ea19c32b554017137d0

### quint8

```
==4361==       Range "quantize_per_tensor, seq = 0"
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
          Range:  100.00%  5.6212ms        20  281.06us  230.17us  352.82us  quantize_per_tensor, seq = 0
 GPU activities:  100.00%  780.85us        10  78.084us  77.633us  78.561us  _ZN2at4cuda75_GLOBAL__N__51_tmpxft_00007fda_00000000_6_affine_quantizer_cpp1_ii_13ee0d7721kernelPointwiseApply2IZZZNS_6native75_GLOBAL__N__51_tmpxft_00007fda_00000000_6_affine_quantizer_cpp1_ii_13ee0d7738quantize_tensor_per_tensor_affine_cudaENS_6TensorES5_dlENKUlvE_clEvENKUlvE0_clEvEUlRfRN3c106quint8EE_fSA_jLi1ELi1ELi1EEEvNS0_6detail10TensorInfoIT0_T2_EENSE_IT1_SG_EESG_T_
      API calls:  100.00%  166.07us        10  16.606us  13.535us  36.578us  cudaLaunchKernel
```

### qint8

```
==12583==       Range "quantize_per_tensor, seq = 0"
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
          Range:  100.00%  5.5765ms        20  278.82us  226.51us  351.23us  quantize_per_tensor, seq = 0
 GPU activities:  100.00%  783.28us        10  78.328us  77.826us  80.386us  _ZN2at4cuda75_GLOBAL__N__51_tmpxft_00007fda_00000000_6_affine_quantizer_cpp1_ii_13ee0d7721kernelPointwiseApply2IZZZNS_6native75_GLOBAL__N__51_tmpxft_00007fda_00000000_6_affine_quantizer_cpp1_ii_13ee0d7738quantize_tensor_per_tensor_affine_cudaENS_6TensorES5_dlENKUlvE_clEvENKUlvE_clEvEUlRfRN3c105qint8EE_fSA_jLi1ELi1ELi1EEEvNS0_6detail10TensorInfoIT0_T2_EENSE_IT1_SG_EESG_T_
      API calls:  100.00%  161.05us        10  16.104us  13.363us  34.284us  cudaLaunchKernel
```

### qint32

```
==17267==       Range "quantize_per_tensor, seq = 0"
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
          Range:  100.00%  19.815ms        20  990.77us  381.03us  1.0717ms  quantize_per_tensor, seq = 0
 GPU activities:  100.00%  1.1778ms        10  117.78us  117.51us  118.44us  _ZN2at4cuda75_GLOBAL__N__51_tmpxft_00007fda_00000000_6_affine_quantizer_cpp1_ii_13ee0d7721kernelPointwiseApply2IZZZNS_6native75_GLOBAL__N__51_tmpxft_00007fda_00000000_6_affine_quantizer_cpp1_ii_13ee0d7738quantize_tensor_per_tensor_affine_cudaENS_6TensorES5_dlENKUlvE_clEvENKUlvE1_clEvEUlRfRN3c106qint32EE_fSA_jLi1ELi1ELi1EEEvNS0_6detail10TensorInfoIT0_T2_EENSE_IT1_SG_EESG_T_
      API calls:  100.00%  172.26us        10  17.226us  14.094us  37.952us  cudaLaunchKernel
```

##

# Environment

```shell
Collecting environment information...
PyTorch version: 1.6.0a0+010771e
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: TITAN V
Nvidia driver version: 440.33.01
cuDNN version: /usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.6.0a0+010771e
[conda] blas                      1.0                         mkl
[conda] magma-cuda102             2.5.2                         1    pytorch
[conda] mkl                       2020.0                      166
[conda] mkl-include               2020.0                      166
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] torch                     1.6.0a0+010771e           dev_0    <develop>
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37312

Differential Revision: D21383938

Pulled By: jerryzh168

fbshipit-source-id: 21539675267c64508a6b9eafcde1a8861d1fb421

Commit:
[FakeLowp] Reset workspace in test (#37799)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37799

Failure to do so will results on some workspace contention.

Test Plan: unittest

Reviewed By: amylittleyang

Differential Revision: D21390900

fbshipit-source-id: 9e837f0f7aae32230740604069308f35b73612b9

zasdfgbnm committed:
Math functions of c10::complex should be overloaded as const reference (#37689)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37689

It has to be this way, otherwise, we will not be able to use it in vec256 because the function pointers declared there are using const reference.

Test Plan: Imported from OSS

Differential Revision: D21394603

Pulled By: anjali411

fbshipit-source-id: daa075b86daaa694489c883d79950a41d6e996ba

jeffdaily committed:
fix ROCm bench CI by increasing first iter timeout (#37633)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37633

Differential Revision: D21395519

Pulled By: ngimel

fbshipit-source-id: 03b31417dde0758db6c189c21b6cb5771c776115

Commit:
Enable HgemmBatched for ROCm (#37483)

Summary:
The purpose of this PR is to enable HgemmBatched for ROCm. Since the inconsistency between CUDA_VERSION and HIP_VERSION, resulting in THCudaBlas_HgemmStridedBatched() not to be called.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37483

Differential Revision: D21395699

Pulled By: ngimel

fbshipit-source-id: c5c22d5f2041d4c9911558b2568fc9ce33ddeb5d

Commit:
[Onnxifi] Cache output shape inference result for OnnxifiOp (#37796)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37796

Shape inference is costly. In bad cases, if we have a lot of uneven tails, we are going to do quite amount of shape inference. This diff will enable each Onnxifi operator to cache the shape inference result for given batch size. In the worst case, we will occupy `num_inference_threads * max_batch_size` OutputReshapeInfo objects per model, where `num_inference_threads` and `max_batch_size` are smaller than 64.

Reviewed By: benjibc

Differential Revision: D21389946

fbshipit-source-id: 23473e64c338d64d15c70292cca0056205d980eb

zasdfgbnm committed:
Remove std::complex to std::complex casting specialization (#37574)

Summary:
This is no longer needed because cuda copy kernel now uses `c10::complex`.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37574

Differential Revision: D21328501

Pulled By: ngimel

fbshipit-source-id: dd5226e8b6c54915fb6ee52240a446f0ca30a800

vkuzo committed:
add qnnpack path for hardtanh (#35779)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/35779

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_hardtanh

Imported from OSS

Differential Revision: D20778588

fbshipit-source-id: 537de42e795a9c67924e1acb1d33b370beb9dbf5

suo committed:
Make a separate cmake option for caffe2 tests (#37721)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37721

Even though we disabled caffe2 test configs in Python, the BUILD_TEST
option was still building caffe2 test cpp binaries and various CI
configurations were running them (since they just run every binary in
`torch/test`).

This PR adds a caffe2-specific BUILD_TEST option (BUILD_CAFFE2_TEST),
which defaults to OFF, and gates the compilation of caffe2 test cpp
binaries under it.

Test Plan: Imported from OSS

Differential Revision: D21369541

Pulled By: suo

fbshipit-source-id: 669cff70c5b53f016e8e016bcb3a99bf3617e1f9

linbinyu committed:
Add several ops for portal NLU/ASR model (again) (#37801)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37801

D21215050  was reverted. Re do it.

Test Plan: build, CI

Reviewed By: iseeyuan

Differential Revision: D21393474

fbshipit-source-id: 2e86d5d1980a122a847e146dc6357627ec31d80d

malfet committed:
[BE] Split pytorch_linux_test into 3 steps (#37808)

Summary:
First one is to download build artifacts
Second is to run tests
Third is to upload test metadata (runs always, even if `Run` step has failed)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37808

Differential Revision: D21398398

Pulled By: malfet

fbshipit-source-id: da23c499a84136e12e88adcc60206ea26bc843c9

jianyuh committed:
[caffe2] L2 regularization for (RowWise)SparseAdagrad fusion on GPUs (#37805)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37805

Resolve the unit test failures after  https://github.com/pytorch/pytorch/pull/37653

Test Plan:
```
buck test mode/dev-nosan //caffe2/caffe2/fb/net_transforms/tests:fuse_sparse_ops_test -- 'test_fuse_sparse_adagrad_with_sparse_lengths_sum_gradient \(caffe2\.caffe2\.fb\.net_transforms\.tests\.fuse_sparse_ops_test\.TestFuseSparseOps\)'
```

```
buck test mode/dev-nosan //caffe2/caffe2/fb/net_transforms/tests:fuse_sparse_ops_test -- 'test_fuse_sparse_adagrad_with_sparse_lengths_weighted_sum_gradient \(caffe2\.caffe2\.fb\.net_transforms\.tests\.fuse_sparse_ops_test\.TestFuseSparseOps\)'
```

Reviewed By: jspark1105

Differential Revision: D21395764

fbshipit-source-id: e8224a1ecbff5dce42ab732c0977de352fe98914

vadimkantorov commented:
If I understand well, this method is only for `numel()==1` tensors? Naming is a bit confusing in this case, since one can expect `torch.zeros(3,4).is_nonzero()` return False, same for some big sparse tensor without any elements (and thus zero by default).

jiayisuse committed:
[TensorPipe/RPC] Serialize and deserialize message (#36197)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/36197

Create APIs to convert between rpc::message and tensorpipe::message
1. tensorpipeSerialize() - converts rpc::message to tensorpipe::message without memory copy (tensors).
2. tensorpipeAllocateMessage - allocates rpc::message based on received tensorpipe descriptor to prepare memory-copy-free receiving.

Test Plan: buck test caffe2/test/cpp/rpc:test_tensorpipe_serialization

Reviewed By: lw

Differential Revision: D20084125

fbshipit-source-id: ffbc310f93443e50261aed752be0fe176610dd2a

jiayisuse committed:
[Tensorpipe/RPC] tensorpipe RPC agent (#35483)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/35483

Implement the initial version of TensorPipe RPC agent, and register to RPC registry to expose to Python interface. As a starter, it utilizes all available TensorPipe transports (shm, uv) and channels (basic, cma).

Test Plan:
https://our.intern.facebook.com/intern/diffusion/FBS/browse/master/fbcode/experimental/jiayisuse/tensorpipe_rpc
  export MASTER_ADDR=127.0.0.1
  export MASTER_PORT=28500
  buck build mode/dev-nosan mode/no-gpu //experimental/jiayisuse/tensorpipe_rpc:main
  ./buck-out/gen/experimental/jiayisuse/tensorpipe_rpc/main.par
  buck build mode/dev-nosan mode/no-gpu //experimental/jiayisuse/tensorpipe_rpc:benchmark
  ./buck-out/gen/experimental/jiayisuse/tensorpipe_rpc/benchmark.par

Multiple connections with async echo
  ./buck-out/gen/experimental/jiayisuse/tensorpipe_rpc/async_echo.par

Reviewed By: lw

Differential Revision: D20088366

fbshipit-source-id: 980f641af3321ca93583c62753e1c9174b7d4afc

nairbv committed:
Revert D21369541: Make a separate cmake option for caffe2 tests

Test Plan: revert-hammer

Differential Revision:
D21369541

Original commit changeset: 669cff70c5b5

fbshipit-source-id: 500d261eaf3f02bcd698d343480b9e951e2844b9

zasdfgbnm committed:
Migrate CUDA fill kernel to c10::complex (#37651)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37651

Test Plan: Imported from OSS

Differential Revision: D21394351

Pulled By: anjali411

fbshipit-source-id: 4fad836700ea25c184dcf4824829f85a0b1e2510

albanD committed:
Fix for hooks with no name (#37785)

Summary:
Fix https://github.com/pytorch/pytorch/issues/37672

Make sure we only access fields that exist and handle python errors correctly.

Before the fix, the given test would throw:
```
AttributeError: 'MyHookClass' object has no attribute '__name__'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_autograd.py", line 432, in test_hook_with_no_name
    x.sum().backward()
  File "/Users/albandes/workspace/pytorch_dev/torch/tensor.py", line 184, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/Users/albandes/workspace/pytorch_dev/torch/autograd/__init__.py", line 115, in backward
    allow_unreachable=True)  # allow_unreachable flag
SystemError: <built-in method run_backward of torch._C._EngineBase object at 0x112fd8100> returned a result with an error set
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37785

Differential Revision: D21387946

Pulled By: albanD

fbshipit-source-id: dcb9afa37b3e10620dc9182d8aa410e7130ffb64

zasdfgbnm committed:
Add vec256 for c10::complex (#37690)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37690

Test Plan: Imported from OSS

Differential Revision: D21394694

Pulled By: anjali411

fbshipit-source-id: 4f0e68280e9c9faf398cfb2d213ecdc4f5cde9fb

jithunnair-amd committed:
Enable test_distributed test test_backend_full_group (#37794)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37794

Differential Revision: D21392441

Pulled By: mrshenli

fbshipit-source-id: 5621b9341a676b695244790ba125d08491a3fe6f

zasdfgbnm committed:
Migrate CUDA unary complex kernel to c10::complex (#37647)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37647

Differential Revision: D21351018

Pulled By: anjali411

fbshipit-source-id: 51e4a4a3bdc9b3f8b9f7a5e0d65c06f209c55401

Commit:
[aten] Pass std::function<> to thread_pool by value, instead of const ref. (#37681)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37681

By passing by value, we can std::move, and avoid unnecessarily copying
args that are part of any std::function/lambda state (e.g. in the jit
interpreter, there is a std::vector<> stack passed in the
InterpreterContinuation)

This makes the api also consistent with e.g. folly and best practices.
Added a minor at::launch() benchmark to test/cpp/, the difference is
mostly noticeable when copying the std::function<> internal args is
non-trivial.

Benchmarks pre/post (min over ~5 runs)
NoData: 5.81 us -> 5.63 us (-3.2%)
WithData(0): 6.67 us -> 5.88 us (-11.8%)
WithData(4): 6.98 us -> 6.51 us (-6.7%)
WithData(256): 9.44 us -> 7.89 (-16.5%)

ghstack-source-id: 103322321

Test Plan:
- perf: buck run mode/opt caffe2/test/cpp/api:parallel_benchmark pre/post
  - correctness buck test mode/dev-nosan caffe2/test/...

Reviewed By: dzhulgakov

Differential Revision: D21355148

fbshipit-source-id: 3567e730845106f1991091e4a892d093e00571c3

cloudhan committed:
Migrate clamp from the TH to Aten (CUDA) (#37646)

Summary:
Fixed https://github.com/pytorch/pytorch/issues/24544

Reference https://github.com/pytorch/pytorch/issues/24507
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37646

Differential Revision: D21395824

Pulled By: VitalyFedyunin

fbshipit-source-id: 111889023d60e3361b5a646bcfb6fb7d5ec969d1

Commit:
[JIT] Fix GetAttr inconsistency (#37424)

Summary:
We were previously only looking at class attributes, so that didn't include methods etc, and would silently give wrong semantics. This makes hasAttr go through the same resolution as our other attribute lookups.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37424

Differential Revision: D21282633

Pulled By: eellison

fbshipit-source-id: 8e970f365c2740d137a02331739c2ed93747b918

kshitij12345 committed:
Migrate `erf` and `erf_` from the TH to Aten (CUDA) : Closes #24558 (#36724)

Summary:
Closes https://github.com/pytorch/pytorch/issues/24558
Benchmark with same build settings on same system.
gcc : version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
CUDA : 10.1
GPU : 1050ti

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.erf(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.erf(a); torch.cuda.synchronize()',
                              setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")',
                              number=t))
```

Before:

```
torch.erf(a) a.numel() == 10000 for 20000 times torch.half
0.29057903600187274
torch.erf(a) a.numel() == 10000 for 20000 times torch.float
0.2836507789979805
torch.erf(a) a.numel() == 10000 for 20000 times torch.double
0.44974555500084534
torch.erf(a) a.numel() == 100000 for 20000 times torch.half
0.31807255600142526
torch.erf(a) a.numel() == 100000 for 20000 times torch.float
0.3216503109979385
torch.erf(a) a.numel() == 100000 for 20000 times torch.double
2.0413486910001666
```

After:

```
torch.erf(a) a.numel() == 10000 for 20000 times torch.half
0.2867302739996376
torch.erf(a) a.numel() == 10000 for 20000 times torch.float
0.28851128199858067
torch.erf(a) a.numel() == 10000 for 20000 times torch.double
0.4592030350013374
torch.erf(a) a.numel() == 100000 for 20000 times torch.half
0.28704102400115517
torch.erf(a) a.numel() == 100000 for 20000 times torch.float
0.29036039400125446
torch.erf(a) a.numel() == 100000 for 20000 times torch.double
2.04035638699861
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/36724

Differential Revision: D21164626

Pulled By: VitalyFedyunin

fbshipit-source-id: e6f3390b2bbb6e8d21e18ffe15f5d49a170fae83

xing-liu committed:
uninitialize output and bag_size in the fast path of EmbeddingBag to save overhead (#36681)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/36681

Test Plan:
Imported from OSS

Unit tests:
python test/run_test.py -i test_nn -- TestNNDeviceTypeCPU.test_EmbeddingBag_per_sample_weights_failures_cpu
python test/run_test.py -i test_nn -- TestNNDeviceTypeCPU.test_EmbeddingBag_per_sample_weights_and_offsets_cpu
python test/run_test.py -i test_nn -- TestNNDeviceTypeCPU.test_EmbeddingBag_per_sample_weights_and_new_offsets_cpu
python test/run_test.py -i test_nn -- TestNNDeviceTypeCPU.test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu
python test/test_nn.py TestNN.test_embeddingbag_from_pretrained
python test/test_nn.py TestNN.test_embeddingbag_from_pretrained_options

Finally run: python test/test_nn.py

Reviewed By: jspark1105

Differential Revision: D21058006

Pulled By: xing-liu

fbshipit-source-id: 65b36a788839e8b722db3e295e58215b5935d6e8

vkuzo committed:
graph mode: more in-place activation handling (#37771)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37771

Adds in place handling for other activations in graph mode

Test Plan:
```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_swap_dequantize
```

Imported from OSS

Differential Revision: D21382825

fbshipit-source-id: 6a4e64bae08fcbfb9bdab92aaac43da98207a1c3

vkuzo committed:
graph mode: round out relu support (#37592)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37592

Makes sure that all the standalone relu flavors are tested in
graph mode.

Test Plan:
```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_swap_dequantize
```

Imported from OSS

Differential Revision: D21366597

fbshipit-source-id: 103848b76a0c65b9adac5bae98b545aa1d30a9e2

xuhdev committed:
Let >> and << support half on CUDA (#37670)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/37670

Differential Revision: D21395325

Pulled By: ngimel

fbshipit-source-id: fcb02f3bee488717cdc1ffc05204970b907d3c3f

kventinel committed:
Fix README for installation from source (#37301)

Summary:
I think, it's help faster compile pytorch from source without errors about incompatible compiler(such as: unsupported GNU version! gcc versions later than 8 are not supported!)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37301

Differential Revision: D21396682

Pulled By: ngimel

fbshipit-source-id: 5e21c36ee550424e820f3aa6e6131ca858994ae4

bharatr21 committed:
Merge from upstream master

Bharat123rox commented:
Made mistakes in resolving conflicts, closing this PR
